features/SPS003_Correct_containment_of_assemblies.feature

```diff
@@ -0,0 +1,11 @@
+@implementer-agreement
+@SPS
+Feature: SPS003 - Correct containment of assemblies
+IfcElement that are aggregated in another IfcElement must not be contained using IfcRelContainedInSpatialStructure.
+
+  Scenario: Agreement on aggregated elements not being contained
+
+      Given A file with Schema Identifier "IFC2x3" or "IFC4" or "IFC4X3_TC1" or "IFC4X3_ADD1" or "IFC4X3"
+      And An IfcElement
+      And Decomposes = not empty
+      Then The value of attribute ContainedInStructure must be empty
```

Comment:
There is the possibility to make them more similar. For example, in the given statements 

```
@then('The value of attribute {attribute} must be {value}')
@then('{attribute} = {value}')
```

This way, as a rule-creator, there's the freedom to define both the given and then statement the same way.
```
      And Decomposes = not empty
      Then ContainedInStructure = empty
```

However, your suggestion of 'is' does not work unfortunately (yet). When the gherkin implementation is decorated with
```
@then('{attribute} is {value}'
```
The rule 'ALB002' 
`Then Each IfcAlignmentHorizontal is nested by a list of only instance(s) of IfcAlignmentSegment`
Will match that the following as the value in the given statement; '`nested by a list of only instance(s) of IfcAlignmentSegment `

Reply:
I fully agree. In the interest of time I'd like to propose to address this at a later stage, because I think it's only a symptom of a deeper problem that:

- because gherkin only matches full sentences we can't really do meaningful composition on a subsentence level. @Ghesselink did nice experiments with using proper grammars instead of just placeholder pattern matching. And ultimately I find that much more promising. Because you define once what a 'value' is (e.g do we use a programming like literal, or something else) we couple it once with a transformer to convert it to a python object and we can use it in all sentences. 
- because of the way of how we log errors and transform applicability state we can't really reuse statements among applicability and requirements. Whereas I would really expect them to be a function `F(instances) -> {True, False}`. When used as an applicability the set of instances gets larger or smaller, when used as a requirement, True triggers a rule_passed event, False triggers an error. This is actually quite easily solvable. We just define our own decorator with the sentence statement as well as an error message around a function returning true or false. And in this decorator we actually delegate to `@then` and `@given`. This would be a great step forward, but the problem with tracking applicability remains. 

I think we've learned a lot and both the features + the implementations are useful. But the matching and reporting provided by gherkin is really a limiting factor, but also not insurmountable, and maybe it also forces us to limit complexity to a level that people can still grok. We just need to have a solid discussion on what direction we want to take and then we can unify the sentences we currently have.